File: day-16-review-day-2/python/set-covering-setup.py
# States we need to cover
states_needed = {"mt", "wa", "or", "id", "nv", "ut", "ca", "az"}

# Radio station and the states they cover
stations = {
  "kone": {"id", "nv", "ut"},
  "ktwo": {"wa", "id", "mt"},
  "kthree": {"or", "nv", "ca"},
  "kfour": {"nv", "ut"},
  "kfive": {"ca", "az"}
}

final_stations = set()

fruits = set(["avocado", "tomato", "banana"])
vegtables = set(["beets", "carrots", "tomato"])
print("Union:", fruits | vegtables)
print("Intersection:", fruits & vegtables)
print("Subtract:", fruits - vegtables)
print("Subtract:", vegtables - fruits)

while states_needed:
  best_station = None
  states_covered = set()
  for station, states in stations.items():
    covered = states_needed & states
    if len(covered) > len(states_covered):
      best_station = station
      states_covered = covered

  states_needed -= states_covered
  final_stations.add(best_station)

print(final_stations)

File: day-16-review-day-2/js/set-covering-setup.js
// States we need to cover (using Set)
const statesNeeded = new Set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"]);

// Radio stations and the states they cover (object with Sets)
const stations = {
  kone: new Set(["id", "nv", "ut"]),
  ktwo: new Set(["wa", "id", "mt"]),
  kthree: new Set(["or", "nv", "ca"]),
  kfour: new Set(["nv", "ut"]),
  kfive: new Set(["ca", "az"]),
};

// Final set of stations
const finalStations = new Set();

function setIntersection(setA, setB) {
  const intersection = new Set();
  for (const item of setA) {
    if (setB.has(item)) {
      intersection.add(item);
    }
  }
  return intersection;
}

function setDifference(setA, setB) {
  const difference = new Set(setA);
  for (const item of setB) {
    difference.delete(item);
  }
  return difference;
}

while (statesNeeded.size > 0) {
  let bestStation = null;
  let statesCovered = new Set(); // States covered by best station
  for (const [station, states] of Object.entries(stations)) {
    const covered = setIntersection(statesNeeded, states);
    if (covered.size > statesCovered.size) {
      bestStation = station;
      statesCovered = covered;
    }
  }
  // Remove covered states from statesNeeded
  for (const state of statesCovered) {
    statesNeeded.delete(state);
  }
  finalStations.add(bestStation);
}

console.log(finalStations);
